### DISCUSSION TITLE
ca65: handling of segments in listing file is confusing and not helpful

### DISCUSSION THREAD

**Discussion by spiro-trikaliotis** [cc65/cc65#1427] [Ideas]
Hello,

I think the handling of segments in ca65 is not optimal and may be
confusing. For this, I created three test files that I attached in a ZIP
in the attachment. Just call make to assemble them and look into the
corresponding listing files (.lst)

ex01.a65 shows "how it just works". I defined a segment "INIT", put my
output there (through .segment), than I put something into DATA, CODE,
and then, again DATA.

Note, however, that there is no hint in the assembler listing into which
segment the data is output. There is only the hint by the .segment
instruction.

Of course, this is enough to find out, isn't it?

It is not.

Look into ex02.a65.

Here, the .segment statement is hidden in a macro. That is, the macro
TEST just changes the segment. You can see the change of the segment
because of the address "000000r" that is printed in the next line, but
you cannot see which segment is the new one.

This could be fixed by having an option that outputs the expanded macro
in the listing file. As far as I see, such an option does not exist,
does it?

Looking into ex03.a65, it get's even more interesting (and confusing):

Here, the macro TEST writes into two different segments, CODE (LDA #0,
A9 00) and DATA (.byte $12).

However, the assembler listing shows:

000001r 1  A9 00 12             TEST

That is, the data that is put into both segments is shown in one line.
The address on the left is completely wrong for the first (!) segment,
it shows the address of the third byte of the line that is put into the
second segment.

You might say that ex03.a65 is an artificial test case, don't you? No,
it is not.

This approach was used by Michael Steil to generate the BASIC variants
at https://www.pagetable.com/?p=46, file token.s: This is an elegant way
to define tables that depend upon each other, in this case, the string
for a BASIC token, the jump address for the function, and even the token
itself!

Having said this, I would like to propose a change of the assembler
output. First of all, data that is put into different segments should be
split up. That is, the line

000001r 1  A9 00 12             TEST

should be split into

00000Ar 1  A9 00                TEST
000001r 1  12

Second, I would like an indicator in which segment it is written. I see
three possibilities:

Variant A:

.segment "CODE"
00000Ar 1  A9 00                TEST
.segment "DATA"
000001r 1  12

Variant B:
CODE:00000Ar 1  A9 00                TEST
DATA:000001r 1  12

Variant C:
segA:00000Ar 1  A9 00                TEST
segB:000001r 1  12

and a hint somewhere that segA is segment "CODE", and segB is segment
"DATA".

What do you think?

I would try to implement this if we agree that this makes sense.

Regards,
Spiro.

**Comment by spiro-trikaliotis** [comment]
Here are the test files mentioned:

[TEST.zip](https://github.com/cc65/cc65/files/6135460/TEST.zip)

**Comment by spiro-trikaliotis** [comment]
And I forgot: When doing conditional compilation, and a .segment statement is in a conditional, it can be hard to find out which segment is currently active. Even in the easy case of ex01.a65, this would be a problem.

**Comment by spiro-trikaliotis** [comment]
And adding even another hint WHY I want this:
I want something similar to #1291: I want to post-process the listing file so it replaces the addresses with the real addresses (and, as far as possible, replace the rr references with the correct values).

I started the project, and I think I come a long way, but the problems mentioned above kill my approach. That's why I want a solution to be able to find out the segment a particular line is assembled into.

**Comment by oliverschmidt** [comment]
First of all, I totally see your point.

I'm coming from a somewhat different perspective. The listing files show the source code (even incl. comments). My naive opinion would have been that macros get expanded in the listing file - but they don't. I consider this a problem.

But there's another point. You are "asking", if it wouldn't in general be better, if one wouldn't need to look at `.segment` directives copied from the source code in order to interpret the listing file. And of course this aspect has a relation to the macro expansion. Because without the need to look at `.segment` directives it's not that important/desirable anymore to have macros expanded.

So my proposal would be to have another or modified listing option that expands macros. As far as I understand that would (without other changes) both

- Make sure that code bytes for different segments never end up in one line.
- Make sure that `.segment` directive are in place.

I would suggest to not add a segment hint/prefix to each line.

**Comment by spiro-trikaliotis** [comment]
I totally agree that an option to have a "complete" listing file (with expanded macros) would be a good option. I was surprised, too, that ca65 does not offer this. However, I am not sure if I could do this addition for now. I am not familiar enough with that part of ca65.

Note, however, this this does **not** solve the problem!

The listing file does not show if a line has been "executed" or not. That is, if I have a _.segment_ instruction inside of a conditional _.if_, then I cannot tell if this _.segment_ takes effect or not.

One could argue that I can see it if the address changes. That's true for most cases, but if the (virtual) PC for two segments is the same (for example, when starting the assembly), this hint is not enough, so, it does not help.

That's why I proposed the option to have the _.segment_ line (variant A). I might not have been specific, but I meant that this line should be added whenever the segment changes, but only then. Thus, the overhead of variant A would be much less than with variant B or C in all practical cases. (Of course, I can construct cases where this is not true, but I consider them artificial).

For my project, adding the addresses to the listing after linking the file, it is essential that I always know exactly to which segment the data is put at a specific line.

**Reply by spiro-trikaliotis** [reply]
or, another option is to make sure to somehow "mark" a line that has an effect (or leave it out).
or, adding an option to only show the lines of the assembly that are taking effect. That is, leave out the lines that are not assembled (for example, because of conditionals)

**Comment by oliverschmidt** [comment]
> The listing file does not show if a line has been "executed" or not. That is, if I have a .segment instruction inside of a conditional .if, then I cannot tell if this .segment takes effect or not.

Hm, maybe I'm missing the point but when I wrote about expanding macros I sort of implied to evaluate conditional disassembly too.

**Reply by spiro-trikaliotis** [reply]
Ok, to me, evaluating conditional assembly was not obvious.
I will have a look if and how I can do it.

**Comment by spiro-trikaliotis** [comment]
I had a first stab in PR #1436.

At first: THIS IS A PREVIEW ONLY, DO NOT APPLY THIS PR!

It is not fully tested yet, and I already know some issues!

I tried to handle the listing file as proposed in #1427. It was more complicated than I thought. ca65 works in a way that it builds the complete listing file in memory, and then outputs it. Alongside this, it generates the tokens of the macro only and processes them instead of the original lines when doing macros. Thus, I had to find a way to match the macro tokens with the source lines.

It works with my test files, but I already have one file where the sync is lost, and the emitted bytes are output to a line "one-off" thant the original source file. I will have a look at it.

However, before I proceed: Is the overall approach acceptable? Am I doing something wrong in a way that this will never be able to be committed?

Additionally: How do I add test cases so that they have to be compared against "known good" files? That is, I do not want to check the output of the assembler, but the output assembly listing file. I found test/ref/, but I do not think it does what I need, does it?

Having some sophisticated macros for testing purposes would be helpful, too. My artificial test cases all work. My cbmrom sources produce a "off-by-one" issue in some cases, which I am currently investigating.

Oh, before I forget, I changed ca65 in the following way:

- Lines that are not processed (because they do not qualify because of conditional assembly, or because they define macros) are not output into the listing file (unless the new option -f, --full is used, see below)
- If -f, -full is used, lines that would be hidden (cf. above) are output as follows: The address is given as "------", and the marker (until now, only "r" for relocatible existed, and space for nothing special) is set to "i".

**Reply by spiro-trikaliotis** [reply]
Oh, and before I forget: Currently, it expands "classic" macros only.
